src/Grpc/Internal/GrpcAsyncCallSpecimenBuilder.cs

```diff
@@ -0,0 +1,122 @@
+
+            var result = method.Invoke(null, new[] { responseStream, callParameters.ResponseHeaders, callParameters.Status, callParameters.Trailers, callParameters.DisposeAction });
+
+            return result;
+        }
+
+        private static object CreateAsyncUnaryCall(Type type, ISpecimenContext context, CallParameters callParameters)
+        {
+            var responseType = type.GenericTypeArguments[0];
+
+            var method = typeof(TestCalls).GetMethod(nameof(TestCalls.AsyncUnaryCall)).MakeGenericMethod(responseType);
+
+            var callResult = context.Resolve(responseType);
+
+            var taskResult = typeof(Task).GetMethod(nameof(Task.FromResult)).MakeGenericMethod(responseType).Invoke(null, new[] { callResult });
+
+            var result = method.Invoke(null, new object[] { taskResult, callParameters.ResponseHeaders, callParameters.Status, callParameters.Trailers, callParameters.DisposeAction });
+
+            return result;
+        }
+
+        private static object CreateAsyncClientStreamingCall(Type type, ISpecimenContext context, CallParameters callParameters)
+        {
+            var requestType = type.GenericTypeArguments[0];
+
+            var responseType = type.GenericTypeArguments[1];
+
+            var method = typeof(TestCalls).GetMethod(nameof(TestCalls.AsyncClientStreamingCall)).MakeGenericMethod(requestType, responseType);
+
+            var requestStream = context.Resolve(typeof(IClientStreamWriter<>).MakeGenericType(requestType));
+
+            var callResult = context.Resolve(responseType);
+
+            var taskResult = typeof(Task).GetMethod(nameof(Task.FromResult)).MakeGenericMethod(responseType).Invoke(null, new[] { callResult });
+
+            var result = method.Invoke(null, new object[] { requestStream, taskResult, callParameters.ResponseHeaders, callParameters.Status, callParameters.Trailers, callParameters.DisposeAction });
+
+            return result;
+        }
+
+        private static object CreateAsyncDuplexStreamingCall(Type type, ISpecimenContext context, CallParameters callParameters)
+        {
+            var requestType = type.GenericTypeArguments[0];
+
+            var responseType = type.GenericTypeArguments[1];
+
+            var method = typeof(TestCalls).GetMethod(nameof(TestCalls.AsyncDuplexStreamingCall)).MakeGenericMethod(requestType, responseType);
+
+            var requestStream = context.Resolve(typeof(IClientStreamWriter<>).MakeGenericType(requestType));
+
+            var responseStream = context.Resolve(typeof(IAsyncStreamReader<>).MakeGenericType(responseType));
+
+            var result = method.Invoke(null, new object[] { requestStream, responseStream, callParameters.ResponseHeaders, callParameters.Status, callParameters.Trailers, callParameters.DisposeAction });
+
+            return result;
+        }
+    }
+
+    public class CallParameters
```

Comment:
I made it into a type because later on I want library users (aka me) to be able to customize these parameters before the call is created. I am not sure how to do that without having a custom attribute per class of tests.

Reply:
@Kralizek do you mean customize as in change the values passed in the `CallParameters` constructor?
This doesn't change how you would populate the instance with data. It only changes how you would create the MethodInfo call parameters
```cs
var parameters = new object[] { requestStream, responseStream }.Concat(callParameters).ToArray();
```
```cs
public class CallParameters : IEnumerable<object>
{
    public CallParameters(Task<Metadata> responseHeaders, Func<Status> status, Func<Metadata> trailers, Action disposeAction)
    {
        ResponseHeaders = responseHeaders;
        Status = status;
        Trailers = trailers;
        DisposeAction = disposeAction;
    }

    public Task<Metadata> ResponseHeaders { get; }

    public Func<Status> Status { get; }

    public Func<Metadata> Trailers { get; }

    public Action DisposeAction { get; }

    public IEnumerator<object> GetEnumerator()
    {
        yield return this.ResponseHeaders;
        yield return this.Status;
        yield return this.Trailers;
        yield return this.DisposeAction;
    }

    IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
}
```

Reply:
oh lol. so lazy :P

Reply:
Here's the lazy evaluated array in case you'll need it. Though the `IEnumerable<>`  implementation should be enough for your needs. You don't enumerate multiple times the parameters instance.

```cs
public class CallParameters : IEnumerable<object>
{
    private readonly Lazy<object[]> lazyArray;

    public CallParameters(Task<Metadata> responseHeaders, Func<Status> status, Func<Metadata> trailers, Action disposeAction)
    {
        ResponseHeaders = responseHeaders;
        Status = status;
        Trailers = trailers;
        DisposeAction = disposeAction;
        lazyArray = new Lazy<object[]>(() => this.ToArray(), true);
    }

    public Task<Metadata> ResponseHeaders { get; }

    public Func<Status> Status { get; }

    public Func<Metadata> Trailers { get; }

    public Action DisposeAction { get; }

    public IEnumerator<object> GetEnumerator()
    {
        yield return this.ResponseHeaders;
        yield return this.Status;
        yield return this.Trailers;
        yield return this.DisposeAction;
    }

    IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();

    public object[] ToArray()
            => this.lazyArray.Value;

    public static implicit operator object[](CallParameters source)
        => source.ToArray();
}
```